### DISCUSSION TITLE
API Script Examples

### DISCUSSION THREAD

**Discussion by markeeisner** [immich-app/immich#7735] [General]
I relatively recently switched to immich from photoprism. I have a custom made digital photo frame running from a raspberry pi and previously manually added photos to a nextcloud folder and used rclone to sync this folder to the pi to update photos.

After moving to immich I wanted to make use of the api to have a "Photoframe" album in immich and have the photos pulled. This would make the process feasible for my wife to manage the photoframe contents with no input from me and without the manual process of populating the nextcloud folder.

I have some basic coding knowledge but most of my experience is from 10 years ago so I was hoping to find some easily adaptable api examples to be able to cobble things together. In the end I couldn't find anything very helpful and nothing immich specific outside of the api documentation.

I've managed to cobble something together in python that works for me, I tried to keep it simple and efficient and easily sharable. With an api key in a file alongside the script, the instance url, album name and destination folder it will download and delete photos from the destination folder only when there are changes.

I wanted to share the script so wondered if there is already a store of these kinds of example scripts for users to take and adapt that I could contribute to? If not what would be the prefered way of sharing this sort of thing?

Some possible options:
- Folder in the repo with scripts and documentation
- Seperate repo with scripts and documentation
- Curated list of scripts in the main repo with links to script authors repos

**Comment by mmomjian** [comment]
How long is the script? Could it be posted as a [guide](https://immich.app/docs/guides/database-gui)?

**Reply by markeeisner** [reply]
65 lines including whitespace so the script and an explaination would probably be well suited to a guide. I'll take a stab at writing it up.

**Comment by soana** [comment]
I'm trying to learn some of this, can you point me to some examples or a place to start?

**Reply by markeeisner** [reply]
My repo for the above mentioned script is here: https://git.orenit.solutions/open/immichalbumpull

A short guide and copy of the script was also merged to the Immich repo but now seems to have been removed.

**Reply by LuucE** [reply]
Hi Mark,

I tried to use your script, but got an TypeError from Python. The error message is:

Traceback (most recent call last):
  File "/home/luuc/immichalbumpull/pull.py", line 37, in <module>
    if item[('album')] == albumname:
       ~~~~^^^^^^^^^^^
TypeError: string indices must be integers, not 'str'

How can I fix this error?

Regards, Luuc.

**Reply by markeeisner** [reply]
Hey Luuc,

Thanks for pointing that out. I hadn't noticed but my photoframe is also giving the exact same error. My setup hasn't had any updates since it was working so I suspect a change in how the Immich API is returning the album list. I will investigate and update the repo when I sort it.

Mark

**Reply by markeeisner** [reply]
@LuucE 

Looks like the API endpoints changed from album -> albums and asset -> assets and I didn't notice the changelog.

https://git.orenit.solutions/open/immichalbumpull is now updated, tested working in my environment.

Mark

**Reply by markeeisner** [reply]
Sorry, looks like the download endpoint is different too. I'll get back to you again when it's actually working!

**Reply by markeeisner** [reply]
@LuucE 

Ok, sorted now. Try again with the latest script and let me know if you have any issues.

**Reply by LuucE** [reply]
Thanks @markeeisner for your quick fix. Now it works as designed.

I wonder if it's possible to use the script for more than one album, maybe all albums? 

Regards, Luuc.

**Reply by markeeisner** [reply]
Hey @LuucE 

It wouldn't be that hard to modify the script for multiple or all albums. Possibly an extra config option or 2 then looping through the album ids to add all the asset references to the main array.

I don't have time to make the modification immediately but I can plan to do it in the next few months.

**Reply by LuucE** [reply]
Hi @markeeisner that would be a nice feature. I can use it to collect all pictures from all albums in Immich, despite who has added them. No problem to wait a few months. 

Regards, Luuc.

**Comment by PureLoneWolf** [comment]
Hi - This script looks like exactly what I need.  I am trying to move over from Google to Immich and this is stopping me.  I use an RClone mount to connect to Google Photos and download a shared album that my family add photos too regularly.

Then I keep that downloaded folder added to my photo frame and it monitors for new photos.

I was trying to figure out if it's possible to Rclone mount in the same way as Google Photos, but can't seem to find a way.  So this script would be perfect.

Unfortunately, I am getting a 404 when I try to connect to the link - Is the script still available by any chance?

Thanks

**Reply by PureLoneWolf** [reply]
Must have been from my side - It seems to be working now and the script works really well, thank you :)

Is there any way to specify multiple albums/targets?  Or would I need to just have multiple copies of the script/config?

**Reply by markeeisner** [reply]
It wasn't you. I'm having an over enthusiastic logging issue with my home gitlab server so it's not up 100% of the time.

I have a request from @LuucE above to be able to select all or multiple albums. I will get around to it within the next few months. For now you can use multiple instances.

**Reply by PureLoneWolf** [reply]
Fantastic - Thank you :)  This genuinely allows me to move away from Google now...

**Comment by markeeisner** [comment]
@PureLoneWolf @LuucE Sorry for the delay. I have updated the script to allow for multiple albums or even all photos on a server.

There are no breaking changes so you can pull the new script from https://git.orenit.solutions/open/immichalbumpull without updating your config. If you want to download multiple albums the album config entry now accepts a comma seperated list of album names. If you leave the album config entry blank it will attempt to download all the images from the server.

I've tested as much as I can. I use my server with a single shared user for both my wife and I so I haven't tested whether it works with shared items. I have assumed, but not tested, that the "all images" option will download all images that the API key has access to whether they are owned by the user or shared.

I have also added some helpful (maybe) console output so you can view the status of the script, how many photos it has found, whether it is downloading or processing a raw file etc. to run the script in cron with only error output I use 'chronic' to silence the script.

**Reply by PureLoneWolf** [reply]
Spot on - Thank you.  I'll test it asap :)

**Comment by markeeisner** [comment]
@PureLoneWolf  @LuucE As the photoframe software I use (https://github.com/helgeerbe/picframe) now supports video I have updated the script to optionally download and use ffmpeg to convert to h264 mp4s for use in the photoframe. https://https://git.orenit.solutions/open/immichalbumpull. There is a new config entry to enable video.
